components/tikv_alloc/src/trace.rs

```diff
@@ -237,6 +237,77 @@ macro_rules! mem_trace {
     }
 }
 
+pub struct MemTraced<T: Default> {
+    item: T,
+    size: usize,
+    node: Option<Arc<dyn MemoryTrace + Send + Sync>>,
+}
+
+impl<T: Default> MemTraced<T> {
+    pub fn new(item: T, size: usize, node: Arc<dyn MemoryTrace + Send + Sync>) -> Self {
+        let node = Some(node);
+        MemTraced { item, size, node }
+    }
+
+    pub fn map<F, U: Default>(mut self, f: F) -> MemTraced<U>
+    where
+        F: FnOnce(T) -> U,
+    {
+        let item = std::mem::take(&mut self.item);
+        MemTraced {
+            item: f(item),
+            size: self.size,
+            node: self.node.take(),
+        }
+    }
+
+    pub fn consume(&mut self) -> T {
```

Comment:
How about using `fn into_inner(self) -> T` as function signature just like [Box::into_inner](https://doc.rust-lang.org/std/boxed/struct.Box.html#method.into_inner) and [Cell::into_inner](https://doc.rust-lang.org/std/cell/struct.Cell.html#method.into_inner)?

Reply:
I prefer this too but It conflicts with `fn drop`. In Rust a struct with explicitly `drop` can't have another method which moves itself.